src/main/kotlin/net/liutikas/sensorsfetch/Fetch.kt

```diff
@@ -51,35 +58,54 @@ private fun fetchDevice(
         successfullyFetchedFiles: MutableList<File>
 ) {
     println("Fetching $sensorName")
-    for (date in endDate..startDate) {
-        print("$date")
-        val fetchedFile = client.fetch(date.toString(), sensorName, outputDirectory)
-        if (fetchedFile == null) {
-            println(" - failure. Failed to fetch ${getUrl(date.toString(), sensorName)}")
-        } else {
-            successfullyFetchedFiles.add(fetchedFile)
-            println(" - success")
+    runBlocking {
+        val deferreds: MutableList<Deferred<FetchResult>> = mutableListOf()
+        for (date in endDate..startDate) {
+            deferreds.add(async {
+                client.fetch(date.toString(), sensorName, outputDirectory)
+            })
+        }
+        deferreds.awaitAll().sortedBy { it.date }.forEach { result ->
+            print(result.date)
+            if (result is SuccessfulResult) {
+                successfullyFetchedFiles.add(result.file)
+                println(" - success")
+            } else {
+                println(" - failure. Failed to fetch ${getUrl(result.date, sensorName)}")
+
+            }
         }
     }
     println("---------------")
 }
 
+private sealed class FetchResult(val date: String)
+private class SuccessfulResult(date: String, val file: File): FetchResult(date)
+private class FailedResult(date: String): FetchResult(date)
+
 private fun getUrl(date: String, sensorName: String): String {
     return "https://archive.sensor.community/${date}/${date}_${sensorName}.csv"
 }
 
-private fun OkHttpClient.fetch(
+private suspend fun OkHttpClient.fetch(
         date: String,
         sensorName: String,
         outputDirectory: File
-): File? {
+): FetchResult {
     val outputFile = File(outputDirectory, "${date}_${sensorName}.csv")
-    if (outputFile.exists()) return outputFile
+    if (outputFile.exists()) return SuccessfulResult(date, outputFile)
     val url = getUrl(date, sensorName)
     val request = Request.Builder().url(url).build()
-    val response = newCall(request).execute()
-    if (response.code != 200) return null
-    val body = response.body?.byteStream() ?: return null
-    Files.copy(body, outputFile.toPath(), StandardCopyOption.REPLACE_EXISTING)
-    return outputFile
+    newCall(request).await().use { response ->
+        if (response.code != 200) return FailedResult(date)
+        val body = response.body?.byteStream() ?: return FailedResult(date)
```

Comment:
You should always call `response.use {}` on the response as its a close able resource.

Reply:
I added .use {} above after the await()